<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Deoxythymidine fragment building &mdash; fluordynamics 0.1.0 documentation</title>
      <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="../_static/jupyter-sphinx.css" type="text/css" />
      <link rel="stylesheet" href="../_static/thebelab.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/custom.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
        <script src="../_static/thebelab-helper.js"></script>
        <script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script src="https://unpkg.com/@jupyter-widgets/html-manager@^0.20.0/dist/embed-amd.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
        <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Nucleic acid coupling chemistries" href="NA_coupling_chemistries.html" />
    <link rel="prev" title="Fragment building" href="../fragment_building.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search"  style="background: #333" >
            <a href="../index.html" class="icon icon-home"> fluordynamics
            <img src="../_static/fluordynamics_logo.png" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Documentation</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="../fluorlabel.html">FluorLabel</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="../fragment_building.html">Fragment building</a><ul class="current">
<li class="toctree-l3 current"><a class="current reference internal" href="#">Deoxythymidine fragment building</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#Build-capped-linker-with-PyMOL">Build capped linker with PyMOL</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Geometry-optimization-and-ESP-calculation-with-Gaussian">Geometry optimization and ESP calculation with Gaussian</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Partial-charge-fitting-with-RESP">Partial charge fitting with RESP</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Fusing-base,-linker-and-fluorophore">Fusing base, linker and fluorophore</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Update-the-force-field">Update the force field</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="NA_coupling_chemistries.html">Nucleic acid coupling chemistries</a></li>
<li class="toctree-l3"><a class="reference internal" href="partial_charges_to_mol2.html">Adding partial charges to mol2 files</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../pymol_plugin.html">PyMOL Plugin</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../fluorburst.html">FluorBurst</a></li>
<li class="toctree-l1"><a class="reference internal" href="../utilities.html">Utilities</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Code</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../module_ff.html">Force field parameterization</a></li>
<li class="toctree-l1"><a class="reference internal" href="../module_fluorbursts.html">Fluorburst</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">References</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../references.html">Publications and related projects</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu"  style="background: #333" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">fluordynamics</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home"></a> &raquo;</li>
          <li><a href="../fluorlabel.html">FluorLabel</a> &raquo;</li>
          <li><a href="../fragment_building.html">Fragment building</a> &raquo;</li>
      <li>Deoxythymidine fragment building</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../_sources/tutorial/deoxythymidine_fragment_building.ipynb.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  
<style>
/* CSS for nbsphinx extension */

/* remove conflicting styling from Sphinx themes */
div.nbinput.container div.prompt *,
div.nboutput.container div.prompt *,
div.nbinput.container div.input_area pre,
div.nboutput.container div.output_area pre,
div.nbinput.container div.input_area .highlight,
div.nboutput.container div.output_area .highlight {
    border: none;
    padding: 0;
    margin: 0;
    box-shadow: none;
}

div.nbinput.container > div[class*=highlight],
div.nboutput.container > div[class*=highlight] {
    margin: 0;
}

div.nbinput.container div.prompt *,
div.nboutput.container div.prompt * {
    background: none;
}

div.nboutput.container div.output_area .highlight,
div.nboutput.container div.output_area pre {
    background: unset;
}

div.nboutput.container div.output_area div.highlight {
    color: unset;  /* override Pygments text color */
}

/* avoid gaps between output lines */
div.nboutput.container div[class*=highlight] pre {
    line-height: normal;
}

/* input/output containers */
div.nbinput.container,
div.nboutput.container {
    display: -webkit-flex;
    display: flex;
    align-items: flex-start;
    margin: 0;
    width: 100%;
}
@media (max-width: 540px) {
    div.nbinput.container,
    div.nboutput.container {
        flex-direction: column;
    }
}

/* input container */
div.nbinput.container {
    padding-top: 5px;
}

/* last container */
div.nblast.container {
    padding-bottom: 5px;
}

/* input prompt */
div.nbinput.container div.prompt pre {
    color: #307FC1;
}

/* output prompt */
div.nboutput.container div.prompt pre {
    color: #BF5B3D;
}

/* all prompts */
div.nbinput.container div.prompt,
div.nboutput.container div.prompt {
    width: 4.5ex;
    padding-top: 5px;
    position: relative;
    user-select: none;
}

div.nbinput.container div.prompt > div,
div.nboutput.container div.prompt > div {
    position: absolute;
    right: 0;
    margin-right: 0.3ex;
}

@media (max-width: 540px) {
    div.nbinput.container div.prompt,
    div.nboutput.container div.prompt {
        width: unset;
        text-align: left;
        padding: 0.4em;
    }
    div.nboutput.container div.prompt.empty {
        padding: 0;
    }

    div.nbinput.container div.prompt > div,
    div.nboutput.container div.prompt > div {
        position: unset;
    }
}

/* disable scrollbars on prompts */
div.nbinput.container div.prompt pre,
div.nboutput.container div.prompt pre {
    overflow: hidden;
}

/* input/output area */
div.nbinput.container div.input_area,
div.nboutput.container div.output_area {
    -webkit-flex: 1;
    flex: 1;
    overflow: auto;
}
@media (max-width: 540px) {
    div.nbinput.container div.input_area,
    div.nboutput.container div.output_area {
        width: 100%;
    }
}

/* input area */
div.nbinput.container div.input_area {
    border: 1px solid #e0e0e0;
    border-radius: 2px;
    /*background: #f5f5f5;*/
}

/* override MathJax center alignment in output cells */
div.nboutput.container div[class*=MathJax] {
    text-align: left !important;
}

/* override sphinx.ext.imgmath center alignment in output cells */
div.nboutput.container div.math p {
    text-align: left;
}

/* standard error */
div.nboutput.container div.output_area.stderr {
    background: #fdd;
}

/* ANSI colors */
.ansi-black-fg { color: #3E424D; }
.ansi-black-bg { background-color: #3E424D; }
.ansi-black-intense-fg { color: #282C36; }
.ansi-black-intense-bg { background-color: #282C36; }
.ansi-red-fg { color: #E75C58; }
.ansi-red-bg { background-color: #E75C58; }
.ansi-red-intense-fg { color: #B22B31; }
.ansi-red-intense-bg { background-color: #B22B31; }
.ansi-green-fg { color: #00A250; }
.ansi-green-bg { background-color: #00A250; }
.ansi-green-intense-fg { color: #007427; }
.ansi-green-intense-bg { background-color: #007427; }
.ansi-yellow-fg { color: #DDB62B; }
.ansi-yellow-bg { background-color: #DDB62B; }
.ansi-yellow-intense-fg { color: #B27D12; }
.ansi-yellow-intense-bg { background-color: #B27D12; }
.ansi-blue-fg { color: #208FFB; }
.ansi-blue-bg { background-color: #208FFB; }
.ansi-blue-intense-fg { color: #0065CA; }
.ansi-blue-intense-bg { background-color: #0065CA; }
.ansi-magenta-fg { color: #D160C4; }
.ansi-magenta-bg { background-color: #D160C4; }
.ansi-magenta-intense-fg { color: #A03196; }
.ansi-magenta-intense-bg { background-color: #A03196; }
.ansi-cyan-fg { color: #60C6C8; }
.ansi-cyan-bg { background-color: #60C6C8; }
.ansi-cyan-intense-fg { color: #258F8F; }
.ansi-cyan-intense-bg { background-color: #258F8F; }
.ansi-white-fg { color: #C5C1B4; }
.ansi-white-bg { background-color: #C5C1B4; }
.ansi-white-intense-fg { color: #A1A6B2; }
.ansi-white-intense-bg { background-color: #A1A6B2; }

.ansi-default-inverse-fg { color: #FFFFFF; }
.ansi-default-inverse-bg { background-color: #000000; }

.ansi-bold { font-weight: bold; }
.ansi-underline { text-decoration: underline; }


div.nbinput.container div.input_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight].math,
div.nboutput.container div.output_area.rendered_html,
div.nboutput.container div.output_area > div.output_javascript,
div.nboutput.container div.output_area:not(.rendered_html) > img{
    padding: 5px;
    margin: 0;
}

/* fix copybtn overflow problem in chromium (needed for 'sphinx_copybutton') */
div.nbinput.container div.input_area > div[class^='highlight'],
div.nboutput.container div.output_area > div[class^='highlight']{
    overflow-y: hidden;
}

/* hide copybtn icon on prompts (needed for 'sphinx_copybutton') */
.prompt a.copybtn {
    display: none;
}

/* Some additional styling taken form the Jupyter notebook CSS */
div.rendered_html table {
  border: none;
  border-collapse: collapse;
  border-spacing: 0;
  color: black;
  font-size: 12px;
  table-layout: fixed;
}
div.rendered_html thead {
  border-bottom: 1px solid black;
  vertical-align: bottom;
}
div.rendered_html tr,
div.rendered_html th,
div.rendered_html td {
  text-align: right;
  vertical-align: middle;
  padding: 0.5em 0.5em;
  line-height: normal;
  white-space: normal;
  max-width: none;
  border: none;
}
div.rendered_html th {
  font-weight: bold;
}
div.rendered_html tbody tr:nth-child(odd) {
  background: #f5f5f5;
}
div.rendered_html tbody tr:hover {
  background: rgba(66, 165, 245, 0.2);
}

/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<section id="Deoxythymidine-fragment-building">
<h1>Deoxythymidine fragment building<a class="headerlink" href="#Deoxythymidine-fragment-building" title="Permalink to this headline"></a></h1>
<p>In this tutorial we will create a new dye-linker-nucleotide fragment which can be used by the PyMOL plugin <strong>FluorLabel</strong> to label nucleic acids for molecular dynamics simulations. Before starting, make sure that you have the following components installed on your machine:</p>
<ul class="simple">
<li><p>PyMOL (install from <a class="reference external" href="https://pymol.org/2/#download">Schrödinger</a>)</p></li>
<li><p>Antechamber (as part of <a class="reference external" href="https://ambermd.org/GetAmber.php#ambertools">AmberTools19</a> or install standalone binary via conda <code class="docutils literal notranslate"><span class="pre">conda</span> <span class="pre">install</span> <span class="pre">ambertools=19</span> <span class="pre">-c</span> <span class="pre">ambermd</span></code> Documentation is available here: <a class="reference external" href="https://ambermd.org/doc12/Amber19.pdf">Amber19 Manual</a></p></li>
<li><p>Acpype (install from <a class="reference external" href="https://alanwilter.github.io/acpype/">Github</a> or via conda <code class="docutils literal notranslate"><span class="pre">conda</span> <span class="pre">install</span> <span class="pre">-c</span> <span class="pre">conda-forge</span> <span class="pre">acpype</span></code></p></li>
<li><p>Pandas and Biopandas (install via conda <code class="docutils literal notranslate"><span class="pre">conda</span> <span class="pre">install</span> <span class="pre">-c</span> <span class="pre">conda-forge</span> <span class="pre">biopandas</span> <span class="pre">pandas</span></code>)</p></li>
<li><p>Fluordynamics</p></li>
</ul>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">from</span> <span class="nn">biopandas.mol2</span> <span class="kn">import</span> <span class="n">PandasMol2</span>
<span class="kn">import</span> <span class="nn">fluordynamics</span> <span class="k">as</span> <span class="nn">fd</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[27]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1">#import importlib</span>
<span class="c1">#importlib.reload(fd.ff)</span>
</pre></div>
</div>
</div>
<p>FluorLabel fuses a fragment that consists of a modified nucleotide and a fluorophore connected through a flexible linker.</p>
<p><img alt="77d24de714da476b9981dcb287346e7f" class="no-scaled-link" src="../_images/dye-linker-base.png" style="width: 500px;" /></p>
<p>A set of commonly used fluorophores has been previously parameterized for Gromacs by the Grubmüller lab (Graen et al. <em>JCTC</em>, <strong>2014</strong>) and distributed through the AMBER-DYES package available on <a class="reference external" href="https://github.com/t-/amber-dyes">Github</a>. A subset of these dyes are located in <code class="docutils literal notranslate"><span class="pre">fragments/dyes/</span></code> (in the Tripos mol2 format). RNA and DNA nucleotides are located in <code class="docutils literal notranslate"><span class="pre">fragments/bases/</span></code>. To fuse the fluorophores to the bases we will need to create linkers which conform to the specific coupling
chemistry used <em>in vitro</em>. This tutorial demonstrates the workflow using a sulfonated Cy3 dye attached to the C5 of deoxythymine as shown above.</p>
<p>The parameterization involves calculating partial charges as well as bond, angle and dihedral parameters for all the atomtypes. The parameters of the bases and the dyes are already available as part of the AMBER and AMBER-DYES force fields. However, those of the linker need to be generated from scratch or in analogy to existing parameters.</p>
<p>We will start by calculating partial charges of the linker by doing a RESP (Restrained ElectroStatic Potential) fit. For this labeling chemistry we use the following 6-carbon linker capped at each both ends with either a methylene or an acetylene (ACE) group to fill the bond valencies and to make the net charge of the spacer equal zero. The capping groups are chosen to best mimic the actual charge distribution when the linker is fused to the dye and the nucleotide.</p>
<p><img alt="f27c4cee0530448a95e97c51f20125f5" class="no-scaled-link" src="../_images/MLE_linker.png" style="width: 450px;" /></p>
<p>The partial charges of the ACE group have been derived when the AMBER-9X force field was designed (Cornell et al. <em>JACS</em> <strong>1995</strong>) and are itself summing up to zero. The methylene charges will be determined as part of the charge fitting procedure with a group charge constraint of -0.15400. This group constraint equals the charge difference between a normal deoxythymine and one where two of the three methyl hydrogens on C5 are removed (partial charge of H71 and H72 is 0.77200 each).</p>
<div class="math notranslate nohighlight">
\[\Delta q = q(\text{DT}_\text{-H71,-H72})-q(\text{DT}_\text{normal}) = -1.15400-(-1) = -0.15400\]</div>
<section id="Build-capped-linker-with-PyMOL">
<h2>Build capped linker with PyMOL<a class="headerlink" href="#Build-capped-linker-with-PyMOL" title="Permalink to this headline"></a></h2>
<p>First, we create the linker with correct bond valencies using the <code class="docutils literal notranslate"><span class="pre">Builder</span></code> tool of PyMOL. Fragments can be added alltogether by <code class="docutils literal notranslate"><span class="pre">Build</span></code>→<code class="docutils literal notranslate"><span class="pre">Fragment</span></code>→<code class="docutils literal notranslate"><span class="pre">Acetylene</span></code>. <img alt="e00f2ace8aba4d228b2eeca218855bf2" class="no-scaled-link" src="tutorial/images/PyMOL_builder.png" style="width: 350px;" /> <img alt="ea7e275e13f94fc483a4cbd9d57f3487" class="no-scaled-link" src="../_images/MLE_linker_PyMOL.png" style="width: 450px;" /></p>
<p>Rename the atoms in a logical way such that there is no overlap with the atom names of the nucleotide (atoms of the linker and the dye can have the same names because there residue name is different). Finally, the linker is saved as a <strong>mol2</strong> file (<code class="docutils literal notranslate"><span class="pre">linker/in/DTM_capped.mol2</span></code>).</p>
<blockquote>
<div><p><strong>Note:</strong> Mol2 files have the advantage that bond valencies are encoded in the format. You can emulate this behavior with PDB files by creating double CONECT entries, however this feature is not recognized by many programs and often leads to error messages instead.</p>
</div></blockquote>
</section>
<section id="Geometry-optimization-and-ESP-calculation-with-Gaussian">
<h2>Geometry optimization and ESP calculation with Gaussian<a class="headerlink" href="#Geometry-optimization-and-ESP-calculation-with-Gaussian" title="Permalink to this headline"></a></h2>
<p>Next, we will do a geometry optimization and calculate an electrostatic potential (ESP) using GAUSSIAN (or alternatively GAMESS). A Gaussian input file can be created via Antechamber:</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>name=MLE_capped
input_folder=&#39;in/&#39;
output_folder=&#39;out/&#39;

antechamber -i &quot;$input_folder/$name&quot;.mol2 -fi mol2 -o &quot;$output_folder/$name&quot;.gin -fo gcrt -gv 1 -ge &quot;$output_folder/$name&quot;.gesp -ch &quot;$output_folder/$name&quot;_opt -nc 0
</pre></div>
</div>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">-i</span></code> input filename</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">-fi</span></code> input file format</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">-o</span></code> output filename</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">-fo</span></code> output file format (gcrt = Gaussian Cartesian)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">-gv</span></code> add keyword to generate gesp file (1 = yes, for Gaussian09)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">ge</span></code> gaussian esp filename generated by iop(6/50=1), default is g09.gesp</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">-ch</span></code> check filename for Gaussian, default is ‘molecule’</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">-nc</span></code> net charge</p></li>
</ul>
<p>To speed up the calculation we will perform the geometry optimization at the B3LYP/6-31G* level of theory followed by the ESP calculation using Hartree Fock (HF/6-31G*). For this purpose, we will slightly modify the Gaussian input file and allow it to run on multiple cores (with the <code class="docutils literal notranslate"><span class="pre">nproc</span></code> keyword).</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>nproc=12

sed &#39;s/#HF.*/\#P b3lyp\/6-31G\* Opt/g&#39; &lt; &quot;$output_folder/$name&quot;.gin | sed &#39;/iop/d&#39; | sed &#39;/.*gesp/d&#39; | sed &quot;/--Link1--/ a %nproc=$nproc&quot; &gt; &quot;$output_folder/$name&quot;_b3lyp_opt.gin

sed &#39;/^[[:space:]]*[A-Z]/d&#39; &lt; &quot;$output_folder/$name&quot;.gin | sed &#39;s/SCF/Geom=check SCF/g&#39;| sed &#39;s/\(\%chk=.*\)opt/\1esp/g&#39; | sed &quot;/--Link1--/ a %nproc=$nproc&quot; &gt; &quot;$output_folder/$name&quot;_hf_esp.gin
</pre></div>
</div>
<p>Run the geometry optimization:</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>g09 &lt; &quot;$output_folder/$name&quot;_b3lyp_opt.gin &gt; &quot;$output_folder/$name&quot;_b3lyp_opt.gout &amp;&amp; cp &quot;$output_folder/$name&quot;_opt.chk &quot;$output_folder/$name&quot;_esp.chk
</pre></div>
</div>
<blockquote>
<div><p><strong>Note:</strong> Make sure that charge and multiplicity are compatible (check log files in case of segmentation errors) files</p>
</div></blockquote>
<p>Calculate the electrostatic surface potential using the checkpoint file of the geometry optimization as start coordinates.</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>g09 &lt; &quot;$output_folder/$name&quot;_hf_esp.gin &gt; &quot;$output_folder/$name&quot;_hf_esp.gout
</pre></div>
</div>
<blockquote>
<div><p><strong>Note:</strong> Remove the coordinates from the original file, in order to use the check point file from the geometry optimization.</p>
</div></blockquote>
</section>
<section id="Partial-charge-fitting-with-RESP">
<h2>Partial charge fitting with RESP<a class="headerlink" href="#Partial-charge-fitting-with-RESP" title="Permalink to this headline"></a></h2>
<p>Use Antechamber to convert the mol2 file of the linker into an Antechamber file (.ac) which can be read by <code class="docutils literal notranslate"><span class="pre">respgen</span></code>.</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>antechamber -i &quot;$input_folder/$name&quot;.mol2 -fi mol2 -o &quot;$output_folder/$name&quot;.ac -fo ac -pf yes -nc 0
</pre></div>
</div>
<p>Create a <code class="docutils literal notranslate"><span class="pre">DTM_capping_group.dat</span></code> file to define which atoms should be assigned a fixed charge and which atoms to group together with a constraint. As mentioned above, we will fix the charges of the ACE atoms according to the AMBER force field and apply a group contraint on the methylene cap.</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>// ACE cap fixed to the charge of AMBER99ff (format: CHARGE &lt;partial_charge&gt; &lt;atom_ID&gt; &lt;atom_name&gt;)
CHARGE 0.5972 10 C16
CHARGE -0.3662 11 C17
CHARGE -0.5679 13 O98
CHARGE 0.1123 32 H97
CHARGE 0.1123 33 H96
CHARGE 0.1123 34 H95

// Group constraint for methylene cap (format: GROUP &lt;number_atoms&gt; &lt;net_charge&gt;)
GROUP 3 -0.15400
ATOM 1 C7
ATOM 16 H01
ATOM 17 H02
</pre></div>
</div>
<ul class="simple">
<li><p><strong>Respgen</strong> generate the input files for two-stage RESP fitting (<code class="docutils literal notranslate"><span class="pre">-f</span></code> specifies the stage (resp1, resp2) and <code class="docutils literal notranslate"><span class="pre">-a</span></code> reads in the frozen and group charges) along with a <code class="docutils literal notranslate"><span class="pre">QIN</span></code> file for the charge constraints.</p></li>
<li><p><strong>Espgen</strong> extracts the ESP from a Gaussian output file.</p></li>
<li><p>The <strong>RESP</strong> fit is performed in two stages: in the first all atoms are allowed to vary, in the second degenerate hydrogens are constrained to have an equal charge. Apart from the input (<code class="docutils literal notranslate"><span class="pre">-i</span></code>) and output (<code class="docutils literal notranslate"><span class="pre">-o</span></code>), additional flags include: <code class="docutils literal notranslate"><span class="pre">-O</span></code> overwrite existing files, <code class="docutils literal notranslate"><span class="pre">-e</span></code> ESP file (input), <code class="docutils literal notranslate"><span class="pre">-q</span></code> charge constraints (input, QIN), <code class="docutils literal notranslate"><span class="pre">-t</span></code> current charges (output), <code class="docutils literal notranslate"><span class="pre">-p</span></code> synopsis of results (output), <code class="docutils literal notranslate"><span class="pre">-s</span></code> ESP for new charges (output).</p></li>
<li><p>Finally, <strong>Antechamber</strong> combines the RESP charges into a new mol2 file (<code class="docutils literal notranslate"><span class="pre">-c</span></code> charge method <em>rc = read in charge</em>, <code class="docutils literal notranslate"><span class="pre">-cf</span></code> charge filename, <code class="docutils literal notranslate"><span class="pre">-pf</span></code> remove intermediary files, <code class="docutils literal notranslate"><span class="pre">-at</span></code> atom type).</p></li>
</ul>
<blockquote>
<div><p><strong>Note:</strong> Link to the <a class="reference external" href="https://upjv.q4md-forcefieldtools.org/RED/resp/">RESP documentation</a></p>
</div></blockquote>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>capping_group=MLE_capping_groups.dat


n_atom=`awk &#39;$1 == &quot;GROUP&quot; {print $2}&#39; &quot;$input_folder/$capping_group&quot;`
group_constraint=`awk &#39;$1 == &quot;GROUP&quot; {print $3}&#39; &quot;$input_folder/$capping_group&quot;`

respgen -i &quot;$output_folder/$name&quot;.ac -o &quot;$output_folder/$name&quot;.respin1 -f resp1 -a &quot;$input_folder/$capping_group&quot;
respgen -i &quot;$output_folder/$name&quot;.ac -o &quot;$output_folder/$name&quot;.respin2 -f resp2 -a &quot;$input_folder/$capping_group&quot;

# since respgen rounds the group constraint to three decimals replace it with the value from the capping group
sed -i &quot;s/$n_atom.*\..*/$n_atom  $group_constraint/g&quot; &quot;$output_folder/$name&quot;.respin1
sed -i &quot;s/$n_atom.*\..*/$n_atom  $group_constraint/g&quot; &quot;$output_folder/$name&quot;.respin2

espgen -i &quot;$output_folder/$name&quot;_hf_esp.gout -o &quot;$output_folder/$name&quot;_hf_esp.esp
mv QIN &quot;$output_folder&quot;/

resp -O -i &quot;$output_folder/$name&quot;.respin1 -o &quot;$output_folder/$name&quot;.respout1 -e &quot;$output_folder/$name&quot;_hf_esp.esp -q  &quot;$output_folder&quot;/QIN -t &quot;$output_folder&quot;/qout_stage1 -p &quot;$output_folder&quot;/punch1 -s &quot;$output_folder&quot;/esout1
resp -O -i &quot;$output_folder/$name&quot;.respin2 -o &quot;$output_folder/$name&quot;.respout2 -e &quot;$output_folder/$name&quot;_hf_esp.esp -q &quot;$output_folder&quot;/qout_stage1 -t &quot;$output_folder&quot;/qout_stage2 -p &quot;$output_folder&quot;/punch2 -s &quot;$output_folder&quot;/esout2

antechamber -i &quot;$output_folder/$name&quot;.ac -fi ac -o &quot;$output_folder/$name&quot;_resp.mol2 -fo mol2 -c rc -cf &quot;$output_folder&quot;/qout_stage2 -pf yes -at amber
</pre></div>
</div>
</section>
<section id="Fusing-base,-linker-and-fluorophore">
<h2>Fusing base, linker and fluorophore<a class="headerlink" href="#Fusing-base,-linker-and-fluorophore" title="Permalink to this headline"></a></h2>
<p>Now that we have all three components (the nucleobase, the linker and the dye all in mol2 format) we can fuse them together using PyMOL. To do this interactively from within this Jupyter notebook, launch a PyMOL server session from a terminal with the following command:</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>pymol -R
</pre></div>
</div>
<p>On a Windows machine you could also create a shortcut that executes the following:</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>C:\path\to\PyMOLWin.exe -R
</pre></div>
</div>
<p>This should open an xml-rpc server running on localhost via port 9123. We connect to this PyMOL instance by:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[19]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">cmd_gui</span> <span class="o">=</span> <span class="n">fd</span><span class="o">.</span><span class="n">jupyter</span><span class="o">.</span><span class="n">connect2pymol</span><span class="p">()</span>
</pre></div>
</div>
</div>
<p>Then, fuse the base and the linker together. Use the <strong>3-Button Editing mode</strong> of PyMOL to rotate the fragments with respect to each other to minimize sterical clashes.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[41]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">names_methylene</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;C7&#39;</span><span class="p">,</span><span class="s1">&#39;H01&#39;</span><span class="p">,</span><span class="s1">&#39;H02&#39;</span><span class="p">]</span>

<span class="n">base_resn</span> <span class="o">=</span> <span class="p">(</span><span class="s1">&#39;deoxythymidine&#39;</span><span class="p">,</span> <span class="s1">&#39;DTM&#39;</span><span class="p">)</span>

<span class="n">cmd_gui</span><span class="o">.</span><span class="n">reinitialize</span><span class="p">()</span>
<span class="n">cmd_gui</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="s1">&#39;../fragments/bases/out/deoxythymidine.mol2&#39;</span><span class="p">)</span>
<span class="n">cmd_gui</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="s1">&#39;../fragments/linkers/MLE/out/MLE_capped_resp.mol2&#39;</span><span class="p">)</span>
<span class="n">cmd_gui</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="s1">&#39;MLE_capped_resp and name </span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="s1">&#39;+&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">names_methylene</span><span class="p">)))</span>
<span class="n">cmd_gui</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="s1">&#39;deoxythymidine and (name H71 or name H72)&#39;</span><span class="p">)</span>
<span class="n">cmd_gui</span><span class="o">.</span><span class="n">fuse</span><span class="p">(</span><span class="s1">&#39;deoxythymidine and name C7&#39;</span><span class="p">,</span> <span class="s1">&#39;MLE_capped_resp and name C8 and resn MLE&#39;</span><span class="p">)</span>
<span class="n">cmd_gui</span><span class="o">.</span><span class="n">delete</span><span class="p">(</span><span class="s1">&#39;deoxythymidine&#39;</span><span class="p">)</span>
<span class="n">cmd_gui</span><span class="o">.</span><span class="n">alter</span><span class="p">(</span><span class="s1">&#39;all&#39;</span><span class="p">,</span> <span class="s1">&#39;type=&quot;ATOM&quot;&#39;</span><span class="p">)</span>
<span class="n">cmd_gui</span><span class="o">.</span><span class="n">alter</span><span class="p">(</span><span class="s1">&#39;all&#39;</span><span class="p">,</span> <span class="s1">&#39;elem=&quot;&quot;&#39;</span><span class="p">)</span> <span class="c1"># PyMOL struggles with atom type definitions in mol2 files, therfore let PyMOL guess the elements itself</span>
<span class="n">cmd_gui</span><span class="o">.</span><span class="n">set_name</span><span class="p">(</span><span class="s1">&#39;MLE_capped_resp&#39;</span><span class="p">,</span> <span class="n">base_resn</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
<span class="n">cmd_gui</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;MLE&#39;</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="n">base_resn</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
</pre></div>
</div>
</div>
<p><img alt="f0ea6bc80a5146e3b1548d6173886512" class="no-scaled-link" src="tutorial/images/PyMOL_threeButtonEditor.png" style="width: 550px;" /></p>
<p>Update the valency of the bond between the linker and the base.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[42]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">cmd_gui</span><span class="o">.</span><span class="n">unbond</span><span class="p">(</span><span class="s1">&#39;resn MLE and name C8&#39;</span><span class="p">,</span> <span class="s1">&#39;resn DT and name C7&#39;</span><span class="p">)</span>
<span class="n">cmd_gui</span><span class="o">.</span><span class="n">bond</span><span class="p">(</span><span class="s1">&#39;resn MLE and name C8&#39;</span><span class="p">,</span> <span class="s1">&#39;resn DT and name C7&#39;</span><span class="p">,</span> <span class="mi">2</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>Save the base-linker as a <strong>mol2</strong> file under <code class="docutils literal notranslate"><span class="pre">fragments/base_linkers/</span></code></p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">save_molecule</span><span class="p">(</span><span class="s1">&#39;../fragments/base_linkers/</span><span class="si">{}</span><span class="s1">.mol2&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">base_resn</span><span class="p">[</span><span class="mi">1</span><span class="p">]),</span><span class="n">base_resn</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="s1">&#39;mol2&#39;</span><span class="p">,</span> <span class="n">overwrite</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>Add the dye to the base-linker and save the whole base-linker-dye fragment as a <strong>pdb</strong> file under <code class="docutils literal notranslate"><span class="pre">fluorlabel/dyes/</span></code> where it can be accessed by the PyMOL plugin FluorLabel.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">dye</span> <span class="o">=</span> <span class="s1">&#39;C3W&#39;</span>
<span class="n">fd</span><span class="o">.</span><span class="n">ff</span><span class="o">.</span><span class="n">couple_dye2baselinker</span><span class="p">(</span><span class="n">dye</span><span class="p">,</span> <span class="n">base_resn</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="p">[</span><span class="s1">&#39;O91&#39;</span><span class="p">,</span> <span class="s1">&#39;C99&#39;</span><span class="p">,</span> <span class="s1">&#39;C27&#39;</span><span class="p">],</span> <span class="p">[</span><span class="s1">&#39;O98&#39;</span><span class="p">,</span> <span class="s1">&#39;C16&#39;</span><span class="p">,</span> <span class="s1">&#39;C17&#39;</span><span class="p">],</span> <span class="p">[</span><span class="s1">&#39;O98&#39;</span><span class="p">,</span> <span class="s1">&#39;C16&#39;</span><span class="p">,</span> <span class="s1">&#39;C17&#39;</span><span class="p">,</span> <span class="s1">&#39;H95&#39;</span><span class="p">,</span> <span class="s1">&#39;H96&#39;</span><span class="p">,</span> <span class="s1">&#39;H97&#39;</span><span class="p">])</span>
<span class="n">save_molecule</span><span class="p">(</span><span class="s1">&#39;../fluorlabel/dyes/</span><span class="si">{}</span><span class="s1">_</span><span class="si">{}</span><span class="s1">.pdb&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">dye</span><span class="p">,</span> <span class="n">base_resn</span><span class="p">[</span><span class="mi">1</span><span class="p">]),</span> <span class="s1">&#39;</span><span class="si">{}</span><span class="s1">_</span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">dye</span><span class="p">,</span> <span class="n">base_resn</span><span class="p">[</span><span class="mi">1</span><span class="p">]),</span> <span class="s1">&#39;pdb&#39;</span><span class="p">,</span> <span class="n">overwrite</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>Finally, add a new entry in the <code class="docutils literal notranslate"><span class="pre">dye_library.json</span></code> file with the filename of your fragment, the dye, the base, the position and the alignment method (base or backbone).</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>[
{&quot;filename&quot;:&quot;C3W_DTM&quot;, &quot;dye&quot;:&quot;sCy3&quot;, &quot;base&quot;:&quot;DT&quot;, &quot;position&quot;:&quot;internal&quot;, &quot;alignment&quot;:&quot;base&quot;}
]
</pre></div>
</div>
</section>
<section id="Update-the-force-field">
<h2>Update the force field<a class="headerlink" href="#Update-the-force-field" title="Permalink to this headline"></a></h2>
<p>We will now create a GROMACS topology (<code class="docutils literal notranslate"><span class="pre">-o</span></code>) of our new base-linker fragment using the AnteChamber PYthon Parser interfacE (Acpype). This time we will choose AMBER as the atom type (<code class="docutils literal notranslate"><span class="pre">-a</span></code>) and read in the previously calculated RESP charges (<code class="docutils literal notranslate"><span class="pre">-c</span></code>). We first need to rename the base (DT) and the linker (MLE) to <strong>DTM</strong> with <code class="docutils literal notranslate"><span class="pre">sed</span></code> because Acpype only accepts one residue name per molecule.</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>cd out/
filename=../../fragments/base_linkers/DTM.mol2
base=DT
linker=MLE

sed &quot;s/${base}1/${base_linker}1/g&quot; &quot;$filename&quot; | sed &quot;s/${linker}1/${base_linker}1/g&quot; &gt; &quot;$base_linker&quot;_ff.mol2
filename=DTM_ff.mol2

acpype -i &quot;$filename&quot; -o gmx -n -1 -a amber -c user
</pre></div>
</div>
<p>Acpype generate an <strong>itp</strong> file as well as force field modification (<strong>frcmod</strong>) file. - The itp file lists all <em>atoms</em>, <em>bonds</em>, <em>angles</em> and <em>dihedrals</em> (both proper and improper) present in the molecule. - The frcmod file lists any <em>atomtypes</em>, <em>bondtypes</em>, <em>angletypes</em> and <em>dihedraltypes</em> which are present in the molecule but missing in the force field (here AMBER because of the <code class="docutils literal notranslate"><span class="pre">-a</span></code> flag in the acpype call).</p>
<blockquote>
<div><p><strong>Note</strong>: The frcmod file is formatted for use with the AMBER MD programs LEaP and Sander. The file and units are converted internally by <code class="docutils literal notranslate"><span class="pre">fd.ff.Parameter.read_frcmod()</span></code> to the GROMACS format.</p>
</div></blockquote>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">fluordyes_itp</span> <span class="o">=</span> <span class="n">fd</span><span class="o">.</span><span class="n">ff</span><span class="o">.</span><span class="n">Molecule</span><span class="o">.</span><span class="n">read_molecule</span><span class="p">(</span><span class="s1">&#39;out/DTM_ff.acpype/DTM_ff_GMX.itp&#39;</span><span class="p">,</span> <span class="s1">&#39;FLUOR-DYNAMICS&#39;</span><span class="p">)</span>
<span class="c1"># change the atom type of O3&#39; from O to OS since this residue is internal and not terminal</span>
<span class="n">fluordyes_itp</span><span class="o">.</span><span class="n">change_type</span><span class="p">(</span><span class="s1">&#39;O3</span><span class="se">\&#39;</span><span class="s1">&#39;</span><span class="p">,</span> <span class="s1">&#39;OS&#39;</span><span class="p">)</span>
<span class="k">for</span> <span class="n">a</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;O98&#39;</span><span class="p">,</span> <span class="s1">&#39;C16&#39;</span><span class="p">,</span> <span class="s1">&#39;C17&#39;</span><span class="p">,</span> <span class="s1">&#39;H95&#39;</span><span class="p">,</span> <span class="s1">&#39;H96&#39;</span><span class="p">,</span> <span class="s1">&#39;H97&#39;</span><span class="p">]:</span>
    <span class="n">fluordyes_itp</span><span class="o">.</span><span class="n">remove_atom</span><span class="p">(</span><span class="n">a</span><span class="p">)</span>
<span class="n">fluordyes_ff</span> <span class="o">=</span> <span class="n">fd</span><span class="o">.</span><span class="n">ff</span><span class="o">.</span><span class="n">Parameters</span><span class="o">.</span><span class="n">read_frcmod</span><span class="p">(</span><span class="s1">&#39;out/DTM_ff.acpype/DTM_ff_AC.frcmod&#39;</span><span class="p">,</span> <span class="n">fluordyes_itp</span><span class="o">.</span><span class="n">atomtypes_molecule</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>Next, we load in the force field parameters from AMBER-DYES</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[18]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">amberdyes_ff</span> <span class="o">=</span> <span class="n">fd</span><span class="o">.</span><span class="n">ff</span><span class="o">.</span><span class="n">Parameters</span><span class="o">.</span><span class="n">read_amberdyes</span><span class="p">([</span><span class="s1">&#39;../forcefields/amberdyes/ffbonded_amberdyes.itp&#39;</span><span class="p">,</span> <span class="s1">&#39;../forcefields/amberdyes/ffnonbonded_amberdyes.itp&#39;</span><span class="p">])</span>
</pre></div>
</div>
</div>
<p>The atoms involved in the bond, angles and dihedrals between linker and fluorophore need to be manually specified because the atom types from AMBER-DYES (for dye) and AMBER (for custom linker) slightly differ. The parameters are defined by analogy to those already present in AMBER-DYES.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[23]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">atoms_amberdyes</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;bondtypes&#39;</span> <span class="p">:</span> <span class="p">[[</span><span class="s1">&#39;ng&#39;</span><span class="p">,</span> <span class="s1">&#39;cg&#39;</span><span class="p">]],</span>
                   <span class="s1">&#39;angletypes&#39;</span><span class="p">:</span> <span class="p">[[</span><span class="s1">&#39;c3g&#39;</span><span class="p">,</span> <span class="s1">&#39;ng&#39;</span><span class="p">,</span> <span class="s1">&#39;cg&#39;</span><span class="p">],</span>
                                  <span class="p">[</span><span class="s1">&#39;hng&#39;</span><span class="p">,</span> <span class="s1">&#39;ng&#39;</span><span class="p">,</span> <span class="s1">&#39;cg&#39;</span><span class="p">],</span>
                                  <span class="p">[</span><span class="s1">&#39;ng&#39;</span><span class="p">,</span> <span class="s1">&#39;cg&#39;</span><span class="p">,</span> <span class="s1">&#39;og&#39;</span><span class="p">],</span>
                                  <span class="p">[</span><span class="s1">&#39;ng&#39;</span><span class="p">,</span> <span class="s1">&#39;cg&#39;</span><span class="p">,</span> <span class="s1">&#39;c3g&#39;</span><span class="p">]],</span>
                   <span class="s1">&#39;propertypes&#39;</span> <span class="p">:</span> <span class="p">[[</span><span class="s1">&#39;c3g&#39;</span><span class="p">,</span> <span class="s1">&#39;c3g&#39;</span><span class="p">,</span> <span class="s1">&#39;cg&#39;</span><span class="p">,</span> <span class="s1">&#39;ng&#39;</span><span class="p">],</span>
                                    <span class="p">[</span><span class="s1">&#39;hcg&#39;</span><span class="p">,</span> <span class="s1">&#39;c3g&#39;</span><span class="p">,</span> <span class="s1">&#39;cg&#39;</span><span class="p">,</span> <span class="s1">&#39;ng&#39;</span><span class="p">],</span>
                                    <span class="p">[</span><span class="s1">&#39;c3g&#39;</span><span class="p">,</span> <span class="s1">&#39;cg&#39;</span><span class="p">,</span> <span class="s1">&#39;ng&#39;</span><span class="p">,</span> <span class="s1">&#39;hng&#39;</span><span class="p">],</span>
                                    <span class="p">[</span><span class="s1">&#39;og&#39;</span><span class="p">,</span> <span class="s1">&#39;cg&#39;</span><span class="p">,</span> <span class="s1">&#39;ng&#39;</span><span class="p">,</span> <span class="s1">&#39;hng&#39;</span><span class="p">],</span>
                                    <span class="p">[</span><span class="s1">&#39;c3g&#39;</span><span class="p">,</span> <span class="s1">&#39;cg&#39;</span><span class="p">,</span> <span class="s1">&#39;ng&#39;</span><span class="p">,</span> <span class="s1">&#39;c3g&#39;</span><span class="p">],</span>
                                    <span class="p">[</span><span class="s1">&#39;og&#39;</span><span class="p">,</span> <span class="s1">&#39;cg&#39;</span><span class="p">,</span> <span class="s1">&#39;ng&#39;</span><span class="p">,</span> <span class="s1">&#39;c3g&#39;</span><span class="p">]]}</span>

<span class="n">atoms_linker</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;bondtypes&#39;</span><span class="p">:</span> <span class="p">[[</span><span class="s1">&#39;N&#39;</span><span class="p">,</span> <span class="s1">&#39;cg&#39;</span><span class="p">]],</span>
                <span class="s1">&#39;angletypes&#39;</span><span class="p">:</span> <span class="p">[[</span><span class="s1">&#39;CT&#39;</span><span class="p">,</span> <span class="s1">&#39;N&#39;</span><span class="p">,</span> <span class="s1">&#39;cg&#39;</span><span class="p">],</span>
                               <span class="p">[</span><span class="s1">&#39;H&#39;</span><span class="p">,</span> <span class="s1">&#39;N&#39;</span><span class="p">,</span> <span class="s1">&#39;cg&#39;</span><span class="p">],</span>
                               <span class="p">[</span><span class="s1">&#39;N&#39;</span><span class="p">,</span> <span class="s1">&#39;cg&#39;</span><span class="p">,</span> <span class="s1">&#39;og&#39;</span><span class="p">],</span>
                               <span class="p">[</span><span class="s1">&#39;N&#39;</span><span class="p">,</span> <span class="s1">&#39;cg&#39;</span><span class="p">,</span> <span class="s1">&#39;c3g&#39;</span><span class="p">]],</span>
                <span class="s1">&#39;propertypes&#39;</span><span class="p">:</span> <span class="p">[[</span><span class="s1">&#39;c3g&#39;</span><span class="p">,</span> <span class="s1">&#39;c3g&#39;</span><span class="p">,</span> <span class="s1">&#39;cg&#39;</span><span class="p">,</span> <span class="s1">&#39;N&#39;</span><span class="p">],</span>
                                <span class="p">[</span><span class="s1">&#39;hcg&#39;</span><span class="p">,</span> <span class="s1">&#39;c3g&#39;</span><span class="p">,</span> <span class="s1">&#39;cg&#39;</span><span class="p">,</span> <span class="s1">&#39;N&#39;</span><span class="p">],</span>
                                <span class="p">[</span><span class="s1">&#39;c3g&#39;</span><span class="p">,</span> <span class="s1">&#39;cg&#39;</span><span class="p">,</span> <span class="s1">&#39;N&#39;</span><span class="p">,</span> <span class="s1">&#39;H&#39;</span><span class="p">],</span>
                                <span class="p">[</span><span class="s1">&#39;og&#39;</span><span class="p">,</span> <span class="s1">&#39;cg&#39;</span><span class="p">,</span> <span class="s1">&#39;N&#39;</span><span class="p">,</span> <span class="s1">&#39;H&#39;</span><span class="p">],</span>
                                <span class="p">[</span><span class="s1">&#39;c3g&#39;</span><span class="p">,</span> <span class="s1">&#39;cg&#39;</span><span class="p">,</span> <span class="s1">&#39;N&#39;</span><span class="p">,</span> <span class="s1">&#39;CT&#39;</span><span class="p">],</span>
                                <span class="p">[</span><span class="s1">&#39;og&#39;</span><span class="p">,</span> <span class="s1">&#39;cg&#39;</span><span class="p">,</span> <span class="s1">&#39;N&#39;</span><span class="p">,</span> <span class="s1">&#39;CT&#39;</span><span class="p">]]}</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[24]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">specialbond_ff</span> <span class="o">=</span> <span class="n">fd</span><span class="o">.</span><span class="n">ff</span><span class="o">.</span><span class="n">Parameters</span><span class="o">.</span><span class="n">read_specialbond</span><span class="p">(</span><span class="n">amberdyes_ff</span><span class="p">,</span> <span class="n">atoms_amberdyes</span><span class="p">,</span> <span class="n">atoms_linker</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>The parameters for the dyes (in <code class="docutils literal notranslate"><span class="pre">amberdyes_ff</span></code>), the linker (in <code class="docutils literal notranslate"><span class="pre">fluordyes_ff</span></code>) and the bond linking the two (in <code class="docutils literal notranslate"><span class="pre">specialbond_ff</span></code>) can now be added to the force field of your choice (here: <code class="docutils literal notranslate"><span class="pre">amber14sb_OL15.ff</span></code>)</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[25]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">fd</span><span class="o">.</span><span class="n">ff</span><span class="o">.</span><span class="n">write_ff</span><span class="p">(</span><span class="s1">&#39;ff14sb&#39;</span><span class="p">,</span> <span class="n">amberdyes_ff</span><span class="p">,</span> <span class="n">fluordyes_ff</span><span class="p">,</span> <span class="n">specialbond_ff</span><span class="p">,</span> <span class="s1">&#39;out/&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>Write a residue topology file (<strong>rtp</strong>) for the linker atoms, bonds and proper dihedrals</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[26]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">fluordyes_itp</span><span class="o">.</span><span class="n">write_rtp</span><span class="p">(</span><span class="s1">&#39;out/DTM.rtp&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>Update the <strong>specbond.dat</strong> file with the bond between the dye and the linker.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">fd</span><span class="o">.</span><span class="n">ff</span><span class="o">.</span><span class="n">update_specbond</span><span class="p">(</span><span class="s1">&#39;C3W C99 1 DTM N99 1 0.160 C3W DTM&#39;</span><span class="p">,</span> <span class="s1">&#39;../forcefields/amberdyes/specbond_amberdyes.dat&#39;</span><span class="p">,</span> <span class="s1">&#39;out/specbond.dat&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>Update the <strong>residuetypes.dat</strong> file with the types of the the dye and the new fragment.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">fd</span><span class="o">.</span><span class="n">ff</span><span class="o">.</span><span class="n">update_residuetypes</span><span class="p">(</span><span class="s1">&#39;C3W DNA&#39;</span><span class="p">,</span> <span class="s1">&#39;../forcefields/amberdyes/residuetypes_amberdyes.dat&#39;</span><span class="p">,</span> <span class="s1">&#39;out/residuetypes.dat&#39;</span><span class="p">,</span> <span class="n">overwrite</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">fd</span><span class="o">.</span><span class="n">ff</span><span class="o">.</span><span class="n">update_residuetypes</span><span class="p">(</span><span class="s1">&#39;DTM DNA&#39;</span><span class="p">,</span> <span class="s1">&#39;out/residuetypes.dat&#39;</span><span class="p">,</span> <span class="s1">&#39;out/residuetypes.dat&#39;</span><span class="p">,</span> <span class="n">overwrite</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="../fragment_building.html" class="btn btn-neutral float-left" title="Fragment building" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="NA_coupling_chemistries.html" class="btn btn-neutral float-right" title="Nucleic acid coupling chemistries" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2020, Fabio D. Steffen, University of Zurich.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>